File: tests/lru_cache_test.cpp
#include "lru_cache.hpp"
#include <cassert>
#include <iostream>
#include <thread>
#include <vector>
int main() {
    LruCache cache(2);
    cache.put("a", "1"); cache.put("b", "2");
    assert(cache.get("a") == "1");
    cache.put("c", "3");
    assert(!cache.get("b"));
    assert(cache.get("a") == "1" && cache.get("c") == "3");
    assert(cache.evictions() == 1);
    cache.put("a", "updated");
    assert(cache.get("a") == "updated");
    assert(cache.erase("a") && !cache.erase("a"));
    LruCache concurrent(100);
    std::vector<std::thread> workers;
    for (int worker = 0; worker < 8; ++worker)
        workers.emplace_back([worker, &concurrent] {
            for (int i = 0; i < 1000; ++i) {
                auto key = std::to_string((worker * 1000 + i) % 150);
                concurrent.put(key, key); concurrent.get(key);
            }
        });
    for (auto& worker : workers) worker.join();
    assert(concurrent.size() <= concurrent.capacity());
    std::cout << "LRU cache tests passed\n";
}

File: src/test/java/com/distributedcache/LruCacheTest.java
package com.distributedcache;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

public final class LruCacheTest {
    private static String text(byte[] value) {
        return new String(value, StandardCharsets.UTF_8);
    }

    public static void main(String[] args) throws InterruptedException {
        LruCache cache = new LruCache(2);
        cache.put("a", "1".getBytes(StandardCharsets.UTF_8));
        cache.put("b", "2".getBytes(StandardCharsets.UTF_8));
        assert text(cache.get("a").orElseThrow()).equals("1");

        cache.put("c", "3".getBytes(StandardCharsets.UTF_8));
        assert cache.get("b").isEmpty() : "b should be the least-recently-used entry";
        assert text(cache.get("a").orElseThrow()).equals("1");
        assert text(cache.get("c").orElseThrow()).equals("3");
        assert cache.stats().evictions() == 1;

        cache.put("a", "updated".getBytes(StandardCharsets.UTF_8));
        assert text(cache.get("a").orElseThrow()).equals("updated");
        assert cache.remove("a");
        assert !cache.remove("a");

        LruCache concurrent = new LruCache(100);
        List<Thread> workers = new ArrayList<>();
        for (int worker = 0; worker < 8; worker++) {
            int workerId = worker;
            workers.add(Thread.startVirtualThread(() -> {
                for (int i = 0; i < 1_000; i++) {
                    String key = Integer.toString((workerId * 1_000 + i) % 150);
                    concurrent.put(key, key.getBytes(StandardCharsets.UTF_8));
                    concurrent.get(key);
                }
            }));
        }
        for (Thread worker : workers) {
            worker.join();
        }
        assert concurrent.stats().size() <= concurrent.stats().capacity();
        System.out.println("Java LRU cache tests passed");
    }
}
